### PR TITLE
Add 'extraQemuOpts' to makeTest

### PR DESCRIPTION
**Pull request by moretea**
##### Motivation for this change

I'm trying to test a piece of software that uses KVM with NixOS' vm test tools. Luckily, KVM supports nested virtualisation [1]. This PR enables me to pass the extra options to QEMU that enable nested VM's.

###### How to test nested KVM support in `makeTest`
1. Ensure that your kernel is configured to supported nested KVM: `boot.extraKernelParams = [ "kvm-intel.nested=1" ];` in `/etc/nixos/configuration.nix`
2. Run the following expression with this git revision of nixpkgs in the `NIX_PATH`
```
{ system ? builtins.currentSystem }:

with import (<nixpkgs> + "/nixos/lib/testing.nix") { inherit system; };
with import (<nixpkgs> + "/nixos/lib/qemu-flags.nix");
with pkgs.lib;

{
  nested_kvm_ok = makeTest {
    name = "nested-kvm-ok";

    extraQemuOpts = "-cpu qemu64,+vmx";

    machine = { config, pkgs, lib, nodes, ... }:
    {
      boot.kernelModules = [ "kvm-intel" ];
      virtualisation.memorySize = 768;
      virtualisation.diskSize = 2048;
    };

    testScript = ''
      startAll;
      $machine->succeed("grep vmx /proc/cpuinfo")
    '';
  };

  nested_kvm_fail = makeTest {
    name = "nested-kvm-fail";

    # Not enabled!
    #extraQemuOpts = "-cpu qemu64,+vmx";

    machine = { config, pkgs, lib, nodes, ... }:
    {
      boot.kernelModules = [ "kvm-intel" ];
      virtualisation.memorySize = 768;
      virtualisation.diskSize = 2048;
    };

    testScript = ''
      startAll;
      $machine->fail("grep vmx /proc/cpuinfo")
    '';
  };
}
```

[1] https://www.kernel.org/doc/Documentation/virtual/kvm/nested-vmx.txt

##### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/lib/testing.nix [hunk 1]
```diff
@@ -114,14 +115,16 @@ rec {
...
+            ${lib.optionalString (extraQemuOpts != null) ''--set QEMU_OPTS "${extraQemuOpts}"''}
           ln -s ${testDriver}/bin/nixos-test-driver $out/bin/nixos-run-vms
           wrapProgram $out/bin/nixos-run-vms \
             --add-flags "$vms" \
             ${lib.optionalString enableOCR "--prefix PATH : '${ocrProg}/bin'"} \
             --set tests 'startAll; joinAll;' \
             --set VLANS '${toString vlans}' \
-            ${lib.optionalString (builtins.length vms == 1) "--set USE_SERIAL 1"}
+            ${lib.optionalString (builtins.length vms == 1) "--set USE_SERIAL 1"} \
+            ${lib.optionalString (extraQemuOpts != null) ''--set QEMU_OPTS "${extraQemuOpts}"''}
```
Comment (Mic92) [inline nixos/lib/testing.nix]: Can extraQemuOpts contain `"`? In this case we should escape them at least.

Comment (moretea) [inline nixos/lib/testing.nix]: Done :)

File: nixos/lib/testing.nix [hunk 2]
```diff
@@ -74,6 +74,7 @@ rec {
     , makeCoverageReport ? false
     , enableOCR ? false
     , name ? "unnamed"
+    , extraQemuOpts ? null # Extra options for QEMU passed to all VM's. String expected
```
Comment (Mic92) [inline nixos/lib/testing.nix]: Would a list works here/make sense? There would be also escapeShellArgs (with an trailing `s`).

**Comment by moretea** [comment]
@Mic92 better :)? Let me this is OK, then I'll flatten all commits into just one commit that you can merge.

**Comment by moretea** [comment]
Apparently, it's already possible to do this by setting `virtualisation.qemu.options` within a VM's config.

```
{ system ? builtins.currentSystem }:

with import (<nixpkgs> + "/nixos/lib/testing.nix") { inherit system; };
with import (<nixpkgs> + "/nixos/lib/qemu-flags.nix");
with pkgs.lib;

{
  nested_kvm_ok = makeTest {
    name = "nested-kvm-ok";

    machine = { config, pkgs, lib, nodes, ... }:
    {
      boot.kernelModules = [ "kvm-intel" ];
      virtualisation.memorySize = 768;
      virtualisation.diskSize = 2048;
      ####### VVV magic is happening there
      virtualisation.qemu.options = ["-cpu qemu64,+vmx"];
    };

    testScript = ''
      startAll;
      $machine->succeed("grep vmx /proc/cpuinfo");
    '';
  };

  nested_kvm_fail = makeTest {
    name = "nested-kvm-ok";
    machine = { config, pkgs, lib, nodes, ... }:
    {
      boot.kernelModules = [ "kvm-intel" ];
      virtualisation.memorySize = 768;
      virtualisation.diskSize = 2048;
    };

    testScript = ''
      startAll;
      $machine->fail("grep vmx /proc/cpuinfo")
    '';
  };
}
```

**Pull request closed without merging**